### PR TITLE
Consent container amp banner

### LINKED ISSUE #1400: Update Cookie Banner Container for AMP
Blocked on https://github.com/bbc/simorgh/issues/1398 & https://github.com/bbc/simorgh/issues/1399

**Is your feature request related to a problem? Please describe.**
Create an AMP container for the Consent banners (Privacy and Cookie) - the logic should be similar to that of the Canonical approach.  We note that there is an AMP implementation of the cookie banner and this was investigated in this proof of concept ticket https://github.com/bbc/simorgh/issues/1297 

**Describe the solution you'd like**
We should have a single container for use by both AMP and Canonical logic. The logic should be in separate files within that container and have the appropriate unit tests. 

**Describe alternatives you've considered**
Having two separate AMP and Canonical containers. But this is contrary to our aim of having containers that can handle both AMP and Canonical logic. 

| Implementation | Agreement Action | Disagree Action |
| ---- | ------------------ | ----------------|
| Privacy Banner AMP | Button | Link with tap action & banner to be hidden & take you to https://www.bbc.co.uk/usingthebbc/your-data-matters | 
| Cookie Banner AMP | Button | Link with tap action & banner to be hidden & take you to https://www.bbc.co.uk/usingthebbc/cookies/how-can-i-change-my-bbc-cookie-settings/ | 

**Testing notes**
[Tester to complete]

Dev insight: Unit tests for props. Cypress test to ensure when buttons are clicked, banners are hidden, etc. 

- [x] Initially labelled with ["Refinement needed"](https://github.com/bbc/simorgh/labels/Refinement%20Needed)

### PR DESCRIPTION
**Pull request by dr3**
Resolves #1400

**Overall change:** Adds privacy and cookie banner for amp

**Code changes:**

- Adds privacy and cookie banner for amp
- Adds a bunch load of tests
- Updates consent banner component to use inheritance due to react restriction

Testing notes: 
The amp banner isnt connected to the canonical one atall, theyre entirely independant.

The amp banner doesnt use cookies, it uses localstorage

Unlike the canonical banner, it doesnt dismiss the privacy banner once youve seen it. You have to click approve on the privacy AND cookie banner in the same page view for them to not return. 

Ask Drew if you have any questions :)

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [x] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/components/ConsentBanner/__snapshots__/index.test.jsx.snap [hunk 1]
```diff
@@ -127,44 +211,103 @@ exports[`ConsentBannerButton should correctly render 1`] = `
...
-  className="c0"
-  onClick={[Function]}
->
-  OK
-</button>
-`;
+@media (min-width:37.5em) {
+  .c2 {
+    width: 22%;
+    margin-right: 3.5%;
```
Comment (pjlee11) [inline src/app/components/ConsentBanner/__snapshots__/index.test.jsx.snap:246]: WOW, if this wasn't copy and past it'd be a no

File: src/app/components/ConsentBanner/index.jsx [hunk 2]
```diff
@@ -113,34 +113,38 @@ export const ConsentBannerText = styled.p`
...
-  ${GEL_GREAT_PRIMER}
-  color: ${C_CONSENT_ACTION};
-  font-weight: 700;
-  background: none;
-  border: none;
-  padding: 0;
-  margin: 0;
-  cursor: pointer;
+// Style `button` and `a` as children due to inability to set `on`
+// prop on styled component as required for the amp useage
```
Comment (pjlee11) [inline src/app/components/ConsentBanner/index.jsx:117]: Nice solution, great comment 😎

File: src/app/containers/ConsentBanner/Banner/index.amp.jsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+const AmpConsentBannerContainer = ({
+  type,
+  acceptAction,
+  rejectAction,
+  promptId,
+  hidden,
+}) => (
+  <ServiceContextConsumer>
+    {({ translations }) => {
+      const messaging = translations.consentBanner[type];
```
Comment (pjlee11) [inline src/app/containers/ConsentBanner/Banner/index.amp.jsx]:
```suggestion
      const consentConfig = translations.consentBanner[type];
```

For the same naming change reasons as the canonical in https://github.com/bbc/simorgh/pull/1428#discussion_r266863365

**Review by pjlee11 — DISMISSED** [review]
Nice work 👍

**Review by ChrisBAshton — COMMENTED** [review]
When I first came to the local AMP page I was seeing the banner despite my cookies being set.

Screen Shot 2019-03-20 at 16 46 55

I clicked on the "OK" button etc and the banner went away and I haven't been able to reproduce it since.

I then cleared my cookies and restarted the server and now I can't see the banner!

Screen Shot 2019-03-20 at 16 51 23

Seems somewhat buggy at the moment! Have you noticed this too?

**Comment by dr3** [comment]
@ChrisBAshton The amp cookie banner doesnt use bbc cookies 🙈 Amp JS controls the storage and uses localstorage, theyre entirely different backends :P

The canonical and amp banner have no affect on eachother

**Review by pjlee11 — DISMISSED** [review]

File: src/app/containers/ConsentBanner/Banner/index.amp.test.jsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import Banner from './index.amp';
+
+describe('Amp Consent Banner Container - Banner', () => {
+  shouldMatchSnapshot(
+    'should correctly render privacy banner',
+    <ServiceContextProvider service="news">
+      <Banner
+        type="privacy"
+        acceptAction={{ tap: [`cookieId.show, privacyId.hide`] }}
+        rejectAction={{ tap: [`cookieId.show, privacyId.hide`] }}
```
Comment (ChrisBAshton) [inline src/app/containers/ConsentBanner/Banner/index.amp.test.jsx]: These could be `"` rather than backticks couldn't they? Just checking why it was done like this?

File: src/app/containers/ConsentBanner/index.amp.jsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+const jsonInlinedScript = data => (
+  <script
+    type="application/json"
+    /* eslint-disable-next-line react/no-danger */
+    dangerouslySetInnerHTML={{ __html: JSON.stringify(data) }}
+  />
+);
+
+// Style `amp-consent` as child due to inability to set
+// `layout` attribute on styled `amp-consent` component
```
Comment (ChrisBAshton) [inline src/app/containers/ConsentBanner/index.amp.jsx:66]: I know you've now done all this work, but want to check if you've explored the styled-components option of passing custom attrs to the DOM...? https://github.com/styled-components/styled-components/issues/1250#issuecomment-377524919

Seems like this would remove the need for the workarounds you've done here. 😄

File: src/app/containers/ConsentBanner/index.amp.jsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    'user-consent': {
+      promptIfUnknownForGeoGroup: 'eu',
+      promptUI: promptId,
+    },
+  },
+};
+
+const jsonInlinedScript = data => (
+  <script
+    type="application/json"
```
Comment (ChrisBAshton) [inline src/app/containers/ConsentBanner/index.amp.jsx:59]: Looks like if we're pulling in react-amphtml we can write this more cleanly: https://www.npmjs.com/package/react-amphtml#target-html

File: src/app/containers/ConsentBanner/index.amp.jsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  <AmpConsentWrapper>
+    <amp-geo layout="nodisplay">{jsonInlinedScript(ampGeoData)}</amp-geo>
+    <amp-consent id={parentId} layout="nodisplay">
+      {jsonInlinedScript(ampConsentData)}
+      <div id={promptId}>
+        <Banner
+          type="privacy"
+          acceptAction={{ tap: [`${cookieId}.show, ${privacyId}.hide`] }}
+          rejectAction={{ tap: [`${cookieId}.show, ${privacyId}.hide`] }}
+          promptId={privacyId}
```
Comment (ChrisBAshton) [inline src/app/containers/ConsentBanner/index.amp.jsx:84]: Shouldn't this have a `hidden` attribute too?

```jsx
hidden={hasUserBeenHereBefore() ? null : true}
```

File: src/app/containers/ConsentBanner/index.amp.test.jsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    act(() => {
+      ReactDOM.render(<AmpContainer />, container);
+    });
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should two banners with correct amp actions and visibility', () => {
```
Comment (ChrisBAshton) [inline src/app/containers/ConsentBanner/index.amp.test.jsx]: Can we reword this to make sense? :)

File: src/app/containers/ConsentBanner/index.amp.test.jsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        hidden: true,
+        promptId: 'cookie',
+        rejectAction: { tap: ['consent.reject'] },
+        type: 'cookie',
+      },
+      {},
+    );
+  });
+
+  it('should render amp-geo element containing inlined JSON', () => {
```
Comment (ChrisBAshton) [inline src/app/containers/ConsentBanner/index.amp.test.jsx]:
```suggestion
  it('should render a single amp-geo element containing inlined JSON', () => {
```

File: src/app/containers/ConsentBanner/index.amp.test.jsx [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    expect(container.querySelectorAll('amp-geo').length).toEqual(1);
+    const ampGeo = container.querySelector('amp-geo');
+    expectNodeToContainInlinedJSON(ampGeo);
+  });
+
+  it('should render amp-consent element containing inlined JSON', () => {
```
Comment (ChrisBAshton) [inline src/app/containers/ConsentBanner/index.amp.test.jsx]:
```suggestion
  it('should render a single amp-consent element containing inlined JSON', () => {
```

(just because that is one of the assertions you're making 👍 )

File: src/server/Document/component.jsx [hunk 11]
```diff
@@ -43,7 +43,19 @@ const Document = ({ assets, app, data, styleTags, helmet }) => {
...
+            <script
+              async
+              custom-element="amp-geo"
+              src="https://cdn.ampproject.org/v0/amp-geo-0.1.js"
+            />
+            <script
+              async
+              custom-element="amp-consent"
+              src="https://cdn.ampproject.org/v0/amp-consent-0.1.js"
+            />
```
Comment (ChrisBAshton) [inline src/server/Document/component.jsx:57]: Is there any way we can conditionally load these only for first time users? (Return users who have their cookies set already would not need to see the banner and wouldn't need the JS)

Micro-optimisation, I know!

**Review by ChrisBAshton — CHANGES_REQUESTED** [review]
Good work, Drew! 👍

File: src/app/containers/ConsentBanner/index.amp.jsx (hunk 5 above)
Comment (dr3) [reply src/app/containers/ConsentBanner/index.amp.jsx:66]: Yeah tried loads, sadly doesnt work, they even have a test for `attrs does not pass non html tags to HTML element` https://github.com/styled-components/styled-components/blob/c3eedbeb79621acd668e15ed2d63107528ed81c3/packages/styled-components/src/test/attrs.test.js#L255-L264

File: src/app/containers/ConsentBanner/index.amp.jsx (hunk 7 above)
Comment (dr3) [reply src/app/containers/ConsentBanner/index.amp.jsx:84]: Nope, amp js does the hiding for us, hidden just tells amp what the expected visual is

File: src/server/Document/component.jsx (hunk 11 above)
Comment (dr3) [reply src/server/Document/component.jsx:57]: Nope, amp doesnt have any client side JS, everything you could ever need needs to be loaded

File: src/app/containers/ConsentBanner/index.amp.test.jsx (hunk 10 above)
Comment (dr3) [reply src/app/containers/ConsentBanner/index.amp.test.jsx]: Done :)

File: src/app/containers/ConsentBanner/index.amp.test.jsx (hunk 9 above)
Comment (dr3) [reply src/app/containers/ConsentBanner/index.amp.test.jsx]: Done :)

File: src/app/containers/ConsentBanner/Banner/index.amp.test.jsx (hunk 4 above)
Comment (dr3) [reply src/app/containers/ConsentBanner/Banner/index.amp.test.jsx]: Ive made them single quotes to match :)

File: src/app/containers/ConsentBanner/index.amp.test.jsx (hunk 8 above)
Comment (dr3) [reply src/app/containers/ConsentBanner/index.amp.test.jsx]: Done :)

File: src/app/containers/ConsentBanner/index.amp.jsx (hunk 6 above)
Comment (dr3) [reply src/app/containers/ConsentBanner/index.amp.jsx:59]: That seems to only be for AmpState
```
<amp-state id="allAnimals">
  <script type="application/json">
    {
      "currentAnimal": "dog"
    }
  </script> 
</amp-state>
```
We need `amp-geo` not `amp-state` which I dont think they do

**Review by pjlee11 — DISMISSED** [review]

File: src/server/Document/component.jsx (hunk 11 above)
Comment (ChrisBAshton) [reply src/server/Document/component.jsx:57]: > amp doesnt have any client side JS

Presumably you mean __`amp-geo`__ doesn't have any client-side JS. AMP itself _is_ client-side JS! 😄

**Review by ChrisBAshton — DISMISSED** [review]

File: src/server/Document/component.jsx (hunk 11 above)
Comment (dr3) [reply src/server/Document/component.jsx:57]: Ah yes soz, I meant amp doesnt have any simorgh client side js :P

**Comment by staylos92** [comment]
The logic looks good to me 👍

Device testing:

Desktop:
Chrome 72.x 👍
Firefox 65.x 👍
Safari 12.x 👍
Edge 👍
IE11 👎 - Privacy and Cookie banners are fixed to the bottom of the screen
IE10 👎- Privacy banner does not show
IE9 👎 - Privacy banner does not show. Images don't appear to load either.

Mobile:
Chrome 72 on Android 👍
Firefox 65 on Android 👍
Opera Mini on Opera Mini 👍
Safari on iOS 12 👍
Safari on iOS 11 👍
Safari on iOS 10 👍
Safari iOS9 👍

**Comment by jamesdonoh** [comment]
Sorry late to the show. Nice implementation, has anyone mentioned to product that the effect of using client storage is that consent never expires, unlike canonical which will expires after one year?

**Comment by dr3** [comment]
I doubt it James, we could raise that, but our hands are pretty tied if they disagree

**Comment by staylos92** [comment]
IE11 now looks good. Thanks Drew 👍

File: src/app/containers/ConsentBanner/index.amp.jsx [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+      'lv',
+      'mt',
+      'nl',
+      'no',
+      'pl',
+      'pt',
+      'ro',
+      'se',
+      'si',
+      'sl',
```
Comment (jamesdonoh) [inline src/app/containers/ConsentBanner/index.amp.jsx:43]: What's the source of this list of country codes? Has it been checked against other lists that we maintain (e.g. in Mozart)? Presumably it will need updating if (let's say) any countries left or joined the EU?

Also, any ideas how we can have confidence that `amp-geo` is doing what we expect it to here?

Comment (dr3) [reply src/app/containers/ConsentBanner/index.amp.jsx:43]: This was copied from https://github.com/bbc/ws-alephamp-queue-renderer/blob/549bace053db39e2902c518089c58179d5328343/src/components/lib/templates/partials/geolocation.mustache

As for the confidence, We have as much confidence as we can. It follows the recomended amp implementation, after that were in the hands of the AMP gods

Comment (dr3) [reply src/app/containers/ConsentBanner/index.amp.jsx:43]: Have raised https://github.com/bbc/simorgh/issues/1465

**Review by pjlee11 — DISMISSED** [review]
Great work!

**Review by ChrisBAshton — APPROVED** [review]

**Review by pjlee11 — APPROVED** [review]
